src/Spice86.Core/Emulator/CPU/CfgCpu/Feeder/InstructionsFeeder.cs

```diff
@@ -37,15 +40,33 @@ public CfgInstruction GetInstructionFromMemory(SegmentedAddress address) {
         // First try to see if it has been encountered before at this address instead of re-parsing.
         // Reason is we don't want several versions of the same instructions hanging around in the graph,
         // this would be bad for successors / predecessors management and self modifying code detection.
-        CfgInstruction? previousMatching = _previousInstructions.GetAtAddressIfMatchesMemory(address);
+        // At this stage, we don't care about non-final fields.
+        // Even if instruction existed before with different non-final field, we create a new instruction.
+        // Reason is we don't want to deal with reducing fields here.
+        CfgInstruction? previousMatching = PreviousInstructions.GetAtAddressIfMatchesMemory(address);
         if (previousMatching != null) {
-            _currentInstructions.SetAsCurrent(previousMatching);
+            CurrentInstructions.SetAsCurrent(previousMatching);
             return previousMatching;
         }
 
+        CfgInstruction parsed = ParseEnsuringUnique(address);
+        CurrentInstructions.SetAsCurrent(parsed);
+        PreviousInstructions.AddInstructionInPrevious(parsed);
+        return parsed;
+    }
+
+    private CfgInstruction ParseEnsuringUnique(SegmentedAddress address) {
         CfgInstruction parsed = _instructionParser.ParseInstructionAt(address);
-        _currentInstructions.SetAsCurrent(parsed);
-        _previousInstructions.AddInstructionInPrevious(parsed);
+        // Let's try with the discriminator reducer to see if the parsed instruction has an existing match
+        HashSet<CfgInstruction>? previousSet = PreviousInstructions.GetAtAddress(address);
+        if (previousSet != null) {
+            foreach (CfgInstruction existing in previousSet) {
```

Comment:
could be a for loop instead of a foreach for more performance.

Reply:
When this piece of code is reached, parser has been called, reducer is being called in a loop.
This will probably not be noticeable since the overhead is way smaller than those 2 other systems.
Fortunately this is not called very often.